samza-core/src/main/java/org/apache/samza/coordinator/metadatastore/CoordinatorStreamStore.java

```diff
@@ -111,6 +111,10 @@ public void init() {
       systemConsumer.start();
       systemProducer.register(SOURCE);
       systemProducer.start();
+      Runtime.getRuntime().addShutdownHook(new Thread(() -> {
```

Comment:
Relying on shutdown hooks for lifecycle management of components is not a good idea. 

The close should be called by whoever created the SystemProducer etc instead. If it's created in this class, this class should have lifecycle methods of it's own, which should be called by the owner of this class.

Reply:
+1, shutdown hooks are being executed when JVM exists, it does not seem to fit this scenario. It could be less appropriate compared to finalize() as the warning messages we get from Kafka is emitted in finalize() as well.

Reply:
@kw2542 We shouldn't use finalize either. The Kafka warning shows that we're leaking client references somewhere without closing them. We should try to find and fix the underlying object lifecycle issue first instead of fixing the symptom.

Reply:
@kw2542 Just FYI, @MabelYC and I were able to identify the places where we're leaking references in JobCoordinator and LocalApplicationRunner.

Reply:
This is awesome!

Reply:
@kw2542 @prateekm  @sborya  Changed to close the clients according to their own lifecycles. Please take a review. Thanks.